typegen/src/utils.rs

```diff
@@ -140,52 +156,58 @@ fn types_equal_inner(
         }
     }
 
-    let a_ty = types.resolve(a).expect("type a should exist in registry");
-    let b_ty = types.resolve(b).expect("type b should exist in registry");
-
     // Paths differ; types won't be equal then!
     if a_ty.path.segments != b_ty.path.segments {
         return false;
     }
 
-    // Names of type params don't line up, so different then!
-    if a_ty
-        .type_params
-        .iter()
-        .zip(&b_ty.type_params)
-        .any(|(a, b)| a.name != b.name)
-    {
-        return false;
-    }
+    #[rustfmt::skip]
+    let mut compare_fields = |
+        a: &Field<PortableForm>,
+        a_params: &GenericsList,
+        b: &Field<PortableForm>,
+        b_params: &GenericsList|
+     -> bool {
+        if a.name != b.name {
+            return false;
+        }
 
-    // We'll lazily extend our type params only if the shapes match.
-    let calc_params = || {
-        let a_params = a_parent_params.extend(&a_ty.type_params);
-        let b_params = b_parent_params.extend(&b_ty.type_params);
-        (a_params, b_params)
+        // Vec<T>, Option<T>, etc and #[scale_info(skip_type_params(T))]
+        let ty_is_skipped_or_wrapped = a_params
+            .index_for_type_id(a.ty.id)
+            .zip(b_params.index_for_type_id(b.ty.id))
+            .is_none();
+
+        // Check that both type names are present or recurse in case of wrapped types
+        a.type_name
```

Comment:
I think this pattern of "zip of Option's` quite hard to read maybe it's just me that reads it as iterators and I wonder whether we could just simplify it by:

```rust
match (a.type_name, b.type_name) {
    (Some(a_type_name), Some(b_type_name)) if ty_is_skipped_or_wrapped => {
        // check that both type names are present in Generic Params and have the same indexes
        a_params
                 .index_for_type_name(a_type_name.as_ref())
                 .zip(b_params.index_for_type_name(b_type_name))
                 .is_some_and(|(a, b)| a == b)
    }
    _ => types_equal_recurse(a.ty.id, a_params, b.ty.id, b_params),
}
```

Anyway either works

Reply:
changed to the suggested variant